components/brave_wallet/browser/zcash/zcash_orchard_storage.cc

```diff
@@ -0,0 +1,445 @@
+/* Copyright (c) 2024 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_wallet/browser/zcash/zcash_orchard_storage.h"
+
+#include <algorithm>
+#include <limits>
+#include <memory>
+#include <string>
+#include <utility>
+
+#include "base/files/file_util.h"
+#include "base/strings/stringprintf.h"
+#include "brave/components/brave_wallet/common/hex_utils.h"
+#include "sql/database.h"
+#include "sql/meta_table.h"
+#include "sql/statement.h"
+#include "sql/transaction.h"
+
+namespace brave_wallet {
+
+namespace {
+#define kNotesTable "notes"
+#define kSpentNotesTable "spent_notes"
+#define kAccountMeta "account_meta"
+
+const int kEmptyDbVersionNumber = 1;
+const int kCurrentVersionNumber = 2;
+
+std::optional<uint32_t> ReadUint32(sql::Statement& statement, size_t index) {
+  auto v = statement.ColumnInt64(index);
+  if (v < 0 || v > std::numeric_limits<uint32_t>::max()) {
+    return std::nullopt;
+  }
+  return static_cast<uint32_t>(v);
+}
+
+}  // namespace
+
+ZCashOrchardStorage::ZCashOrchardStorage(base::FilePath path_to_database)
+    : db_file_path_(std::move(path_to_database)) {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
+  database_ = std::make_unique<sql::Database>();
+}
+
+ZCashOrchardStorage::~ZCashOrchardStorage() {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
+}
+
+bool ZCashOrchardStorage::EnsureDbInit() {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
+  if (database_->is_open()) {
+    return true;
+  }
+  return CreateOrUpdateDatabase();
+}
+
+void ZCashOrchardStorage::ResetDatabase() {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
+
+  database_->Close();
+  sql::Database::Delete(db_file_path_);
+}
+
+bool ZCashOrchardStorage::CreateOrUpdateDatabase() {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
+
+  const base::FilePath dir = db_file_path_.DirName();
+  if (!base::DirectoryExists(dir) && !base::CreateDirectory(dir)) {
+    NOTREACHED_IN_MIGRATION();
```

Comment:
Occurences of `NOTREACHED_IN_MIGRATION()` in this file are not _NOTREACHED_ but just runtime failures
All should be removed

Reply:
Done